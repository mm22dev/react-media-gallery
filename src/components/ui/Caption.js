import React from 'react'
import PropTypes from 'prop-types'
import { connectWithPopupSettings } from '../../providers/PopupSettings'
import { Fade } from '../_common'

const CaptionUI = ({ fade, style, children, popupSettings }) => {
  const { usePopupDefaultStyling, popupCaptionSelector, popupCaptionHeight, popupPadding, popupCaptionTextSize, popupBlankCaptionText } = popupSettings

  const defaultStyle = {
    ...(usePopupDefaultStyling && {
      ...(popupCaptionTextSize && { fontSize: popupCaptionTextSize }),
      position: 'absolute',
      left: 0,
      bottom: 0,
      width: '100%',
      textAlign: 'center',
      height: `${popupCaptionHeight}px`,
      lineHeight: `${popupCaptionHeight}px`,
      whiteSpace: 'nowrap',
      overflow: 'hidden',
      textOverflow: 'ellipsis',
      padding: `0 ${popupPadding ? `${popupPadding}px` : 0}`
    })
  }

  const captionContent = !children || children.length === 0 ? popupBlankCaptionText : children

  return (
    // PopupSettings provider sets popupCaptionSelector to null if usePopupCaption is false
    popupCaptionSelector
      ? <Fade {...fade}>
        <div className={popupCaptionSelector.slice(1)} style={{ ...defaultStyle, ...style }} title={captionContent}>
          { captionContent }
        </div>
      </Fade>
      : null
  )
}

CaptionUI.propTypes = {
  fade: PropTypes.object,
  style: PropTypes.object,
  children: PropTypes.string,
  popupSettings: PropTypes.object
}

CaptionUI.defaultProps = {
  fade: { inProp: false, duration: 0, immediate: true },
  style: {},
  children: ''
}

const Caption = connectWithPopupSettings(CaptionUI)
Caption.displayName = 'Caption'
export default Caption
