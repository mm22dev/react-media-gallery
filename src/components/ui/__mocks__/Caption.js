const Caption = () => <div className='caption caption-mock'></div>

Caption.displayName = 'CaptionMock'

export default Caption
