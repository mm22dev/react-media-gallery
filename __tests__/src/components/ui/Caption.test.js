import Caption from '../../../../src/components/ui/Caption'
import toJSON from 'enzyme-to-json'
import { compose } from 'redux'
import defaultPopupSettings from '../../../../src/data/popupSettings.json'

const { shallow, mount } = Enzyme

describe('<Caption /> Component', () => {
  let { popupSettings } = defaultPopupSettings
  popupSettings.usePopupCaption = true

  describe('Rendering UI', () => {
    it('Renders correctly', () => {
      compose(expect, toJSON, shallow)(<Caption />).toMatchSnapshot()
    })

    it('Does not Apply default style if required', () => {
      const fadeInjectedStyle = { display: 'block', opacity: 0 }
      expect(
        mount(<Caption popupSettings={{ ...popupSettings, usePopupDefaultStyling: false }} />)
          .find(popupSettings.popupCaptionSelector)
          .get(0)
          .props
          .style
      ).toEqual(fadeInjectedStyle)
    })
  })

  describe('Applying default properties', () => {
    it('Assigns correct default content', () => {
      expect(mount(<Caption popupSettings={popupSettings} />).childAt(0).text()).toBe('(untitled)')
    })

    it('Assigns correct default content', () => {
      expect(mount(<Caption popupSettings={popupSettings} children={''} />).childAt(0).text()).toBe('(untitled)')
    })
  })

  describe('Receiving props from parent.', () => {
    it('Assigns correct custom text size', () => {
      expect(
        mount(<Caption popupSettings={{ ...popupSettings, popupCaptionTextSize: '1em' }} />)
          .find(popupSettings.popupCaptionSelector)
          .get(0)
          .props
          .style
          .fontSize
      ).toBe('1em')
    })

    it('Fills the content with proper text', () => {
      expect(mount(<Caption popupSettings={popupSettings} children={'mock text'} />).childAt(0).text()).toBe('mock text')
    })
  })
})
