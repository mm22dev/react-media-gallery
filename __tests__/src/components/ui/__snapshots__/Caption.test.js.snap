// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Caption /> Component Rendering UI Renders correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
